stm32U59x and stm32U5Ax series supporting the USB OTG HS

FRASTM opened the pull request:
Add the support of the USB OTG HS node to the stm32U59x/5Ax/5Fx/5Gx  devices
Enable the OTG HS node on the stm32u5a9j_dk  disco kit

Adapt the usb device driver to power and clock the OTG HS and PHY correctly

Requires https://github.com/zephyrproject-rtos/zephyr/pull/76115

FRASTM committed:
dts: arm: stm32u59x serie with OTG HS instance

Add the USB OTG HS node for the stm32U59x/5Ax/5Fx/5Gx devices

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: st: stm32u5a9 disco kit has a OTG HS peripheral

Enable the USB OTG HS node of the stm32u5a9j_dk board

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: usb: stm32 usb device controller clock and power

Special sequence to enable clock and power for the OTG HS
peripheral of the stm32U59x serie
This code is based on the stm32Cube HAL_HCD_MspInit/DeInit.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: udc: stm32 usb device controller clock and power

Special sequence to enable clock and power for the OTG HS
peripheral of the stm32U59x serie
This code is based on the stm32Cube HAL_HCD_MspInit/DeInit.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
west build -p always -b stm32u5a9j_dk samples/subsys/usb/cdc_acm -- -DCONF_FILE=usbd_next_prj.conf

```
*** Booting Zephyr OS build v3.7.0-rc2-460-g59129e816840 ***
[00:00:00.010,000] <inf> cdc_acm_echo: USB device support enabled
[00:00:00.010,000] <inf> cdc_acm_echo: Wait for DTR
[00:00:00.015,000] <inf> cdc_acm_echo: USBD message: Device suspended
[00:00:03.620,000] <wrn> udc: Spurious suspend/resume event
[00:00:03.620,000] <inf> cdc_acm_echo: USBD message: Device suspended
[00:00:04.021,000] <inf> cdc_acm_echo: USBD message: Bus reset
[00:00:04.101,000] <inf> cdc_acm_echo: USBD message: Bus reset
[00:00:04.198,000] <inf> cdc_acm_echo: USBD message: CDC ACM line coding
[00:00:04.198,000] <inf> cdc_acm_echo: Baudrate 9600
```

FRASTM commented:
west build  -p always -b stm32u5a9j_dk tests/subsys/usb/device/
```
TESTSUITE device_usb succeeded                                                                                    
                                                                                                                  
------ TESTSUITE SUMMARY START ------                                                                             
                                                                                                                  
SUITE PASS - 100.00% [device_usb]: pass = 5, fail = 0, skip = 0, total = 5 duration = 0.032 seconds               
 - PASS - [device_usb.test_usb_dc_api] duration = 0.001 seconds                                                   
 - PASS - [device_usb.test_usb_dc_api_invalid] duration = 0.019 seconds                                           
 - PASS - [device_usb.test_usb_dc_api_read_write] duration = 0.006 seconds                                        
 - PASS - [device_usb.test_usb_deconfig] duration = 0.002 seconds                                                 
 - PASS - [device_usb.test_usb_disable] duration = 0.004 seconds                                                  
                                                                                                                  
------ TESTSUITE SUMMARY END ------                                                                               
                                                                                                                  
===================================================================                                               
PROJECT EXECUTION SUCCESSFUL  
```

FRASTM committed:
drivers: usb: stm32 usb device controller correct USB PHY

Set the correct the phy_itface depending on the setting
Save few lines after resetting the priv->pcd structure
with the memset

Signed-off-by: Francois Ramu <francois.ramu@st.com>

ydamigos commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -962,13 +966,36 @@ static int priv_clock_enable(void)
+		LL_PWR_EnableVddUSB();
+
+		/* Configure VOSR register of USB HSTranceiverSupply(); */
+		LL_PWR_EnableUSBPowerSupply();
```
RM0456 mentions:
```
On STM32U59x/5Ax/5Fx/5Gx devices only:
a) Make sure the voltage scaling is in range 1 or in range 2 (using VOS[1:0]
in PWR_VOSR).
b) Make sure the EPOD booster clock is enabled (using PLL1MBOOST[3:0]
in RCC_PLL1CFGR).
```
Should we check them before setting USBPWREN and USBBOOSTEN in PWR_VOSR?

FRASTM replied:
The voltage scaling is set when configuring the stm32u5 sys clock. Yes, we can add a simple check on LL_PWR_REGU_VOLTAGE_SCALE1 or LL_PWR_REGU_VOLTAGE_SCALE2 and return an error if not correct.
The USB EPOD  booster is enabled by the LL_PWR_EnableUSBEPODBooster and yes we can add a loop waiting for the USBBOOSTRDY flag (LL_PWR_IsActiveFlag_USBBOOST )

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -217,14 +217,35 @@ static int usb_dc_stm32_clock_enable(void)
-#if defined(PWR_USBSCR_USB33SV) || defined(PWR_SVMCR_USV)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) && defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* Sequence to enable the power of the OTG HS on a stm32U5 serie : Enable VDDUSB */
+	if (__HAL_RCC_PWR_IS_CLK_DISABLED()) {
+		__HAL_RCC_PWR_CLK_ENABLE();
```
For consistency, can we use LL API in stead of __HAL_RCC macros ?

FRASTM replied:
yes, possible with LL_AHB3_GRP1_IsEnabledClock()

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -217,14 +217,35 @@ static int usb_dc_stm32_clock_enable(void)
+		LL_PWR_EnableUSBEPODBooster();
+	}
+
+	__HAL_RCC_SYSCFG_CLK_ENABLE();
```
Ditto

FRASTM replied:
possible with LL_APB3_GRP1_EnableClock

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -217,14 +217,35 @@ static int usb_dc_stm32_clock_enable(void)
+	}
+
+	__HAL_RCC_SYSCFG_CLK_ENABLE();
+	/* Set the OTG PHY reference clock selection */
+	HAL_SYSCFG_SetOTGPHYReferenceClockSelection(SYSCFG_OTG_HS_PHY_CLK_SELECT_1);
+	/* Configuring the SYSCFG registers OTG_HS PHY : OTG_HS PHY enable*/
+	HAL_SYSCFG_EnableOTGPHY(SYSCFG_OTG_HS_PHY_ENABLE);
```
Ditto if LL API available.

FRASTM replied:
no LL driver  for SYSCFG

erwango commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -277,6 +298,9 @@ static int usb_dc_stm32_clock_enable(void)
 #if defined(CONFIG_SOC_SERIES_STM32H7X)
 	LL_AHB1_GRP1_DisableClockSleep(LL_AHB1_GRP1_PERIPH_USB1OTGHSULPI);
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
+	LL_AHB2_GRP1_DisableClockStopSleep(LL_AHB2_GRP1_PERIPH_OTG_HS ||
```
Disabling `LL_AHB2_GRP1_PERIPH_OTG_HS` is not stricly in line with this section covered by comment: "Disable ULPI interface (for external high-speed PHY) clock in sleep/low-power mode".
I'd suggest to split and add a specific comment.

FRASTM replied:
put in another commit  3689885f5ec4b5cc8f2db3d4916b8787bb9bd130 and

FRASTM replied:
LL_AHB2_GRP1_DisableClockStopSleep must disable sleep of the OTG AND PHY  clock  at the same time, else the usb connection is not detected

erwango commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -966,13 +966,36 @@ static int priv_clock_enable(void)
-#if defined(PWR_USBSCR_USB33SV) || defined(PWR_SVMCR_USV)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) && defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* Sequence to enable the power of the OTG HS on a stm32U5 serie : Enable VDDUSB */
+	if (__HAL_RCC_PWR_IS_CLK_DISABLED()) {
```
See comments on the other driver

erwango reviewed (changes requested):
Good job.
While at it, can you check if this could fix https://github.com/zephyrproject-rtos/zephyr/discussions/64895 ?

FRASTM committed:
drivers: udc: stm32 usb device controller in sleep mode

Force disabling the USB OTG HS and PHY clock during sleepmode
By default, that clock is enabled by clock gating during sleep
mode. Like stm32H7, it has to be kept running.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: usb: stm32 usb device controller in sleep mode

Force disabling the USB OTG HS and PHY clock during sleepmode
By default, that clock is enabled by clock gating during sleep
mode. Like stm32H7, it has to be kept running.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
- rebase on 4ddd3ebd51fb85390c34bc580487f43d1144cd24
- new commit for disabling sleep mode (for each usb stm32 driver)
- add a check on Voltage scaling range 
- wait for USB Epod Boost ready after setting
